### PR TITLE
SOV.approveAndCall

### PR DESCRIPTION
**Pull request by jameshowlett977**
- SOV is ERC20 with an additional function: approveAndCall
- ApprovalReceiver: base implementation of a contract which need approval
- inherited contracts: RSOV, Staking, Vesting, DevelopmentVesting
- FeeSharingProxy is an exception here since this contract needs approval for different tokens and uses SOV only for getting voting power of the user

### CODE REVIEW

**Review by ororopickpocket — APPROVED** [review]
nice work!

**Comment by ororopickpocket** [comment]
I requested a review from evgeny as well

File: contracts/governance/Staking/Staking.sol [hunk 1]
```diff
@@ -20,14 +20,44 @@ contract Staking is IStaking, WeightedStaking {
 		address stakeFor,
 		address delegatee
 	) public {
+		_stake(msg.sender, amount, until, stakeFor, delegatee);
```
Comment (eMarchenko) [inline contracts/governance/Staking/Staking.sol:23]: mark `stake` as external and call `_stake` internally instead

File: contracts/governance/ApprovalReceiver.sol [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	/**
+	 * @notice receives approval from SOV token
+	 * @param _data the data will be used for low level call
+	 */
+	function receiveApproval(
+		address, /*_sender*/
+		uint256, /*_amount*/
+		address, /*_token*/
+		bytes memory _data
+	) public {
```
Comment (eMarchenko) [inline contracts/governance/ApprovalReceiver.sol:25]: `external` - it saves a bit of gas and improves readability

File: contracts/governance/ApprovalReceiver.sol [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+	/**
+	 * @notice receives approval from SOV token
+	 * @param _data the data will be used for low level call
+	 */
+	function receiveApproval(
+		address, /*_sender*/
+		uint256, /*_amount*/
+		address, /*_token*/
+		bytes memory _data
```
Comment (eMarchenko) [inline contracts/governance/ApprovalReceiver.sol:24]: all targets receive `_sender, _amount` as first two arguments. You can check that `_data[4:36] == _sender && _data[36:68] == _amount` (python notation for slices)
You also can check if `_token == msg.sender == _getToken()`

**Review by eMarchenko — CHANGES_REQUESTED** [review]
- please check if the `Staking._stake` is working correctly
- personally I don't like the way `receiveApproval` is working (external call to self), but it is subjective
- there are several suggestions (minor improvements)

all in all the PR is fine

File: contracts/governance/Staking/Staking.sol [hunk 4]
```diff
@@ -39,7 +69,7 @@ contract Staking is IStaking, WeightedStaking {
 
 		uint96 previousBalance = currentBalance(stakeFor, until);
 		//increase stake
-		_increaseStake(amount, stakeFor, until);
+		_increaseStake(sender, amount, stakeFor, until);
 
 		if (previousBalance == 0) {
 			//regular delegation if it's a first stake
```
[the remark this answers is not in the document]
Comment (jameshowlett977) [reply contracts/governance/Staking/Staking.sol:75]: you're right, fixed

File: contracts/governance/Vesting/DevelopmentVesting.sol [hunk 5]
```diff
@@ -85,15 +86,30 @@ contract DevelopmentVesting is Ownable {
 	 * @param _amount the amount of tokens to send
 	 */
 	function depositTokens(uint256 _amount) public {
+		_depositTokens(msg.sender, _amount);
+	}
+
+	/**
+	 * @notice deposit tokens to this contract, these tokens can be withdrawn any time by an owner
+	 * @dev this function will be invoked from receiveApproval
+	 * @dev SOV.approveAndCall -> this.receiveApproval -> this.mintWithApproval
```
[the remark this answers is not in the document]
Comment (jameshowlett977) [reply contracts/governance/Vesting/DevelopmentVesting.sol]: done

File: contracts/governance/Staking/Staking.sol (hunk 1 above)
Comment (jameshowlett977) [reply contracts/governance/Staking/Staking.sol:23]: done

File: contracts/governance/ApprovalReceiver.sol (hunk 2 above)
Comment (jameshowlett977) [reply contracts/governance/ApprovalReceiver.sol]: done

File: contracts/governance/ApprovalReceiver.sol (hunk 3 above)
Comment (jameshowlett977) [reply contracts/governance/ApprovalReceiver.sol]: done

File: contracts/governance/ApprovalReceiver.sol [hunk 6]
```diff
@@ -18,12 +18,39 @@ contract ApprovalReceiver is ErrorDecoder, IApproveAndCall {
...
+				break;
+			}
+		}
+		require(isAllowed, "method is not allowed");
+
+		//check sender and amount
+		bytes32 dummy;
+		address sender;
+		uint256 amount;
+		(dummy, sender, amount) = abi.decode(
```
Comment (eMarchenko) [inline contracts/governance/ApprovalReceiver.sol:46]: jfyi: `(, address sender, uint256 amount) = abi.decode(...` - a more compact notation - I've skipped `dummy` and linlined variable declaration

**Review by eMarchenko — APPROVED** [review]

File: contracts/governance/ApprovalReceiver.sol (hunk 6 above)
Comment (jameshowlett977) [reply contracts/governance/ApprovalReceiver.sol:46]: agree :-)

**Pull request merged**